docs/bytecode.md

```diff
@@ -0,0 +1,208 @@
+the `strings` command can be used to see what attributes were missed:
+
+_Note: The following trace is from a previous revision where the `scatter` attribute was not
+implemented. Currently all types/attrs are implemented and log only shows 
+the dialect name `stablehlo` and the custom `stablehlo.frontend_attributes` and `stablehlo.sharding` properties._
+
+```
+$ stablehlo-opt -emit-bytecode file.mlir | strings | grep stablehlo
+#stablehlo.scatter<update_window_dims = [1], inserted_window_dims = [0, 1], scatter_dims_to_operand_dims = [0, 1], index_vector_dim = 1>
+stablehlo
+stablehlo.frontend_attributes
+stablehlo.sharding
+```
+
+### Debugging Bytecode with Traces
+
+Each read/write function called during bytecoding is traced, and can be viewed using the flag: `-debug-only=stablehlo-bytecode`.
+
+```
+stablehlo-opt -emit-bytecode -debug-only=stablehlo-bytecode ../tmp.mlir
+Called: writeType(mlir::Type, mlir::DialectBytecodeWriter &)::(anonymous class)::operator()(auto) const [type:auto = mlir::stablehlo::TokenType]
+Called: writeAttribute(mlir::Attribute, mlir::DialectBytecodeWriter &)::(anonymous class)::operator()(auto) const [attr:auto = mlir::stablehlo::TransposeAttr]
+Called: writeAttribute(mlir::Attribute, mlir::DialectBytecodeWriter &)::(anonymous class)::operator()(auto) const [attr:auto = mlir::stablehlo::RngAlgorithmAttr]
+Called: writeAttribute(mlir::Attribute, mlir::DialectBytecodeWriter &)::(anonymous class)::operator()(auto) const [attr:auto = mlir::stablehlo::ChannelHandleAttr]
+Called: writeAttribute(mlir::Attribute, mlir::DialectBytecodeWriter &)::(anonymous class)::operator()(auto) const [attr:auto = mlir::stablehlo::ChannelHandleAttr]
+Called: writeAttribute(mlir::Attribute, mlir::DialectBytecodeWriter &)::(anonymous class)::operator()(auto) const [attr:auto = mlir::stablehlo::TypeExtensionsAttr]
+...
+
+stablehlo-opt -emit-bytecode -debug-only=stablehlo-bytecode bytecoded_file.mlir
+Called: readComparisonDirectionAttr(mlir::DialectBytecodeReader &) const
+Called: readTypeExtensionsAttr(mlir::DialectBytecodeReader &) const
+Called: readChannelHandleAttr(mlir::DialectBytecodeReader &) const
+Called: readChannelHandleAttr(mlir::DialectBytecodeReader &) const
+Called: readRngAlgorithmAttr(mlir::DialectBytecodeReader &) const
+```
+
+### Adding Bytecode for a New Type / Attribute
+
+Adding bytecode for a new type or attribute is simple. In the file 
+`StablehloBytecode.cpp` search for the term `TO ADD ATTRIBUTE` or `TO ADD TYPE`
+depending on the change. Ensure that each location tagged with `TO ADD` 
+instructions is addressed. If so, bytecode for the attr/type should be generated
+on next call to `stablehlo-opt -emit-bytecode`.
+
+
+### Encoding `enum class` values
+Enum class values can be encoded as their underlying numeric types using `varint`. Currently all enums in StableHLO use `uint32_t` as the underlying value.
+
+## Open Questions
+### Is the compatibility guarantee of StableHLO dependent on other dialects?
```

Comment:
Yes and no, but more yes it is dependent. I recommend to Eugene that an audit be done of all the attribute types used. This is also why for this layer one should either use things that aspire to be stable (as StableHLO does) such as ml_program (e.g., use ml_program.func rather than func.func) or are stable by "simplicity" (e.g., IntegerAttr) and curate. Now even with that, there are still levels of freedom to provide guarantees here independent of this, but starting with not overcomplicating ones life is good.

Reply:
I suggested before using some verifier (possibly a pass) to check the content of a module and have an allowlist of type/attributes that can be present.

Reply:
Created #69